platform/security/certificate/security-crl-generator/src/main/java/org/codice/ddf/security/crl/generator/CrlGenerator.java

```diff
@@ -0,0 +1,355 @@
+  private static final int INITIAL_DELAY = 0;
+  private static final int SCHEDULER_INTERVAL = 30;
+  private static final int FILE_BACKED_STREAM_THRESHOLD = 1000000;
+  static final String CRL_PROPERTY_KEY = Merlin.OLD_PREFIX + Merlin.X509_CRL_FILE;
+  static final String DEM_CRL = "/localCrl.crl";
+  static final String PEM_CRL = "/localCrl.pem";
+
+  @VisibleForTesting
+  static String issuerEncryptionPropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/issuer/encryption.properties").getPath();
+
+  @VisibleForTesting
+  static String issuerSignaturePropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/issuer/signature.properties").getPath();
+
+  @VisibleForTesting
+  static String serverEncryptionPropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/server/encryption.properties").getPath();
+
+  @VisibleForTesting
+  static String serverSignaturePropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/server/signature.properties").getPath();
+
+  @VisibleForTesting
+  static String crlFileLocation = new AbsolutePathResolver("etc/keystores").getPath();
+
+  private final ClientFactoryFactory factory;
+  private final EventAdmin eventAdmin;
+  private final ScheduledExecutorService scheduler;
+  private String crlLocationUrl;
+  private boolean crlByUrlEnabled;
+  private ScheduledFuture<?> handle;
+
+  public CrlGenerator(ClientFactoryFactory factory, EventAdmin eventAdmin) {
+    this.factory = factory;
+    this.eventAdmin = eventAdmin;
+    this.scheduler =
+        Executors.newSingleThreadScheduledExecutor(
+            StandardThreadFactoryBuilder.newThreadFactory("crlThread"));
+  }
+  /** Pulls down the CRL file provided via URL and writes it locally. */
+  @Override
+  public synchronized void run() {
+    if (!crlByUrlEnabled || crlLocationUrl == null) {
+      return;
+    }
+    if (!crlLocationUrl.startsWith(HTTPS)) {
+      postErrorEvent("The provided URL was not an HTTPS URL.");
+      return;
+    }
+    Object entity = getRemoteCrl();
+    if (!(entity instanceof InputStream)) {
+      postErrorEvent("Unable to download the CRL.");
+      return;
+    }
+    FileBackedOutputStream fileBackedOutputStream = null;
+    try {
+      // Read the response content and get the byte source
+      ByteSource byteSource;
+      try (InputStream entityStream = (InputStream) entity) {
+        fileBackedOutputStream = new FileBackedOutputStream(FILE_BACKED_STREAM_THRESHOLD);
+        IOUtils.copy(entityStream, fileBackedOutputStream);
+        fileBackedOutputStream.close();
+        byteSource = fileBackedOutputStream.asByteSource();
+      }
+      File crlFile = getCrlFile(byteSource);
+      // Verify that the CRL is valid
+      if (!crlIsValid(byteSource)) {
+        postErrorEvent("An error occurred while validating the CRL.");
+        return;
+      }
+      writeCrlToFile(byteSource, crlFile);
+    } catch (IOException e) {
```

Comment:
We let the current thread finish before we schedule a new task. Does this still apply in that case?

Reply:
Yes because the set method will cancel the future and that will result in the thread being interrupted and system I/O will throw back an InterruptedIOException and abort when they detect this situation.

Reply:
Correct but when we're canceling it, we let the current task finish. We don't interrupt it.

Reply:
That is done automatically when you cancel the future

Reply:
Yeah that's why I do `handle.cancel(false);` instead

Reply:
The code still shows `cancel(true)`

Reply:
https://github.com/codice/ddf/pull/4059/files#diff-cf6db63ca090fabcfeb6af02d45a5aefR317

Reply:
It wasn't when I last looked at it 😄 ... all good then, you can ignore my comments for all `InterruptedIOException`, although some of them won't break anything anyways